csharp/src/Apache.Arrow/ArrowBuffer.Builder.cs

```diff
@@ -109,20 +109,19 @@ public Builder<T> Clear()
                 return this;
             }
 
-            public ArrowBuffer Build(MemoryPool pool = default)
+            public ArrowBuffer Build(MemoryAllocator allocator = default)
             {
-                int length;
-                checked
-                {
-                    length = (int)BitUtility.RoundUpToMultipleOf64(_buffer.Length);
-                }
+                int length = checked((int)BitUtility.RoundUpToMultipleOf64(_buffer.Length));
 
-                var memoryPool = pool ?? MemoryPool.Default.Value;
-                var memory = memoryPool.Allocate(length);
+                var memoryAllocator = allocator ?? MemoryAllocator.Default.Value;
+                var memoryOwner = memoryAllocator.Allocate(length);
 
-                Memory.CopyTo(memory);
+                if (memoryOwner != null)
```

Comment:
`memoryOwner` will be `null` when you pass in a `length == 0` to `memoryAllocator.Allocate(length)`.

Looking at this code more - it is using `_buffer.Length` which isn't correct. It should be using `Length`, which can be different (for example if I used the builder to create a `1000` elements, and then called `Clear()` and added `10` elements, I don't need a buffer big enough for `1000` elements.

Reply:
Fixing this is a little bigger than I thought it would be. So I separated it into a commit on its own.

Reply:
Thank you for the explanation, it was not obvious to me.

Have you considered using the Null Object Pattern and returning an empty memory instead of null in such case?

Reply:
Thanks for the suggestion, @adamsitnik. I've opened https://issues.apache.org/jira/browse/ARROW-5277 to implement this.